ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util.hs

```diff
@@ -374,3 +388,48 @@ fib n = round $ phi ** fromIntegral n / sq5
 eitherToMaybe :: Either a b -> Maybe b
 eitherToMaybe (Left _)  = Nothing
 eitherToMaybe (Right x) = Just x
+
+{-------------------------------------------------------------------------------
+  Fuse code, i.e. regions of code that will throw an exception if accessed
+  concurrently.
+-------------------------------------------------------------------------------}
+
+-- | A WithIntactFuse expression is one of which a single instance can be
+-- evaluated at any given moment in the execution of the program.
+newtype WithIntactFuse m a = WithIntactFuse (m a)
+  deriving newtype (Functor, Applicative, Monad, MonadThrow, MonadCatch)
+
+instance MonadTrans WithIntactFuse where
+  lift = WithIntactFuse
+
+-- | Declare that an action is not safe to be run concurrently, i.e. two
+-- concurrent accesses should throw an exception.
+blowFuse = WithIntactFuse
+blowFuse :: m a -> WithIntactFuse m a
+
+-- | Lock required to run a 'WithIntactFuse'
+data Fuse m = Fuse String (MVar m ())
+
+newFuse :: MonadMVar m => String -> m (Fuse m)
+newFuse name = Fuse name <$> newMVar ()
+
+-- | Run the 'WithIntactFuse' action throwing an exception if it is already
+-- running in a separate thread.
+blowAndResetFuse ::
+     (MonadThrow m, MonadMVar m)
+  => Fuse m
+  -> WithIntactFuse m a
+  -> m a
+blowAndResetFuse (Fuse name m) (WithIntactFuse io) = do
+    () <- onlyOne =<< tryTakeMVar m
```

Comment:
```suggestion
    onlyOne =<< tryTakeMVar m
```
(if this statement would return something other than `()`, we would get a `-Wunused-do-bind` warning)

Reply:
Done